public/app/features/correlations/CorrelationsPage.tsx

```diff
@@ -0,0 +1,215 @@
+  const { remove, get } = useCorrelations();
+
+  useEffect(() => {
+    get.execute();
+    // we only want to fetch data on first render
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const canWriteCorrelations = contextSrv.hasPermission(AccessControlAction.DataSourcesWrite);
+
+  const handleAdd = useCallback(() => {
+    get.execute();
+    setIsAdding(false);
+  }, [get]);
+
+  const handleUpdate = useCallback(() => {
+    get.execute();
+  }, [get]);
+
+  const handleRemove = useCallback<(params: RemoveCorrelationParams) => void>(
+    async (correlation) => {
+      await remove.execute(correlation);
+      get.execute();
+    },
+    [remove, get]
+  );
+
+  const RowActions = useCallback(
+    ({
+      row: {
+        original: {
+          source: { uid: sourceUID, readOnly },
+          uid,
+        },
+      },
+    }: CellProps<CorrelationData, void>) =>
+      !readOnly && (
+        <DeleteButton
+          aria-label="delete correlation"
+          onConfirm={() => handleRemove({ sourceUID, uid })}
+          closeOnConfirm
+        />
+      ),
+    [handleRemove]
+  );
+
+  const columns = useMemo<Array<Column<CorrelationData>>>(
+    () => [
+      {
+        cell: InfoCell,
+        shrink: true,
+        visible: (data) => data.some(isSourceReadOnly),
+      },
+      {
+        id: 'source',
+        header: 'Source',
+        cell: DataSourceCell,
+        sortType: sortDatasource,
+      },
+      {
+        id: 'target',
+        header: 'Target',
+        cell: DataSourceCell,
+        sortType: sortDatasource,
+      },
+      { id: 'label', header: 'Label', sortType: 'alphanumeric' },
+      {
+        cell: RowActions,
+        shrink: true,
+        visible: (data) => canWriteCorrelations && data.some(negate(isSourceReadOnly)),
+      },
+    ],
+    [RowActions, canWriteCorrelations]
+  );
+
+  const data = useMemo(() => get.value, [get.value]);
+
+  const showEmptyListCTA = data?.length === 0 && !isAdding && (!get.error || get.error.status === 404);
+
+  return (
+    <Page navModel={navModel}>
+      <Page.Contents>
+        <div>
+          <HorizontalGroup justify="space-between">
+            <div>
+              <h4>Correlations</h4>
+              <p>Define how data living in different data sources relates to each other.</p>
+            </div>
+            {canWriteCorrelations && data?.length !== 0 && data !== undefined && !isAdding && (
+              <Button icon="plus" onClick={() => setIsAdding(true)}>
+                Add new
+              </Button>
+            )}
+          </HorizontalGroup>
+        </div>
+
+        {!data && get.loading && (
+          <div className={loaderWrapper}>
+            <LoadingPlaceholder text="loading..." />
+          </div>
+        )}
+
+        {showEmptyListCTA && <EmptyCorrelationsCTA onClick={() => setIsAdding(true)} />}
+
+        {
+          // This error is not actionable, it'd be nice to have a recovery button
```

Comment:
Is this worthy to be solved right here or in another PR? Would you like to create an issue for it?

Reply:
https://github.com/grafana/grafana/issues/54288